python/ql/lib/semmle/python/dataflow/new/internal/DataFlowDispatch.qll

```diff
@@ -0,0 +1,1540 @@
+  not result.(ParameterNodeImpl).isParameterOf(_, any(ParameterPosition pp | pp.isSelf()))
+}
+
+/**
+ * Gets a reference to the result of calling `super` without any argument, where the
+ * call happened in the method `func` (either a method or a classmethod).
+ */
+Node superCallNoArgumentTracker(Function func) {
+  superCallNoArgumentTracker(TypeTracker::end(), func).flowsTo(result)
+}
+
+/**
+ * Gets a reference to the result of calling `super` with 2 arguments, where the
+ * first is a reference to the class `cls`, and the second argument is `obj`.
+ */
+private TypeTrackingNode superCallTwoArgumentTracker(TypeTracker t, Class cls, Node obj) {
+  t.start() and
+  exists(CallCfgNode call | result = call |
+    call = getSuperCall() and
+    call.getArg(0) = classTracker(cls) and
+    call.getArg(1) = obj
+  )
+  or
+  exists(TypeTracker t2 | result = superCallTwoArgumentTracker(t2, cls, obj).track(t2, t)) and
+  not result.(ParameterNodeImpl).isParameterOf(_, any(ParameterPosition pp | pp.isSelf()))
+}
+
+/**
+ * Gets a reference to the result of calling `super` with 2 arguments, where the
+ * first is a reference to the class `cls`, and the second argument is `obj`.
+ */
+Node superCallTwoArgumentTracker(Class cls, Node obj) {
+  superCallTwoArgumentTracker(TypeTracker::end(), cls, obj).flowsTo(result)
+}
+
+// =============================================================================
+// MRO
+// =============================================================================
+/**
+ * Gets a direct superclass of the argument `cls`, if any.
+ *
+ * For `A` with the class definition `class A(B, C)` it will have results `B` and `C`.
+ */
+Class getADirectSuperclass(Class cls) { cls.getABase() = classTracker(result).asExpr() }
+
+/**
+ * Gets a direct subclass of the argument `cls`, if any.
+ *
+ *For `B` with the class definition `class A(B)` it will have result `A`.
+ */
+Class getADirectSubclass(Class cls) { cls = getADirectSuperclass(result) }
+
+/**
+ * Gets a class that, from an approximated MRO calculation, might be the next class used
+ * for member-lookup when `super().attr` is used inside the class `cls`.
+ *
+ * In the example below, with `cls=B`, this predicate will have `A` and `C` as results.
+ * ```py
+ * class A: pass
+ * class B(A): pass
+ * class C(A): pass
+ * class D(B, C): pass
+ * ```
+ *
+ * NOTE: This approximation does not handle all cases correctly, and in the example
+ * below, with `cls=A` will not have any results, although it should include `Y`.
+ *
+ * ```py
+ * class A: pass
+ * class B(A): pass
+ * class X: pass
+ * class Y(X): pass
+ * class Ex(B, Y): pass
+ * ```
+ *
+ * NOTE for debugging the results of this predicate: Since a class can be part of
+ * multiple MROs, results from this predicate might only be valid in some, but not all,
+ * inheritance chains (such as the result `C` for `cls=B` in the first example -- this
+ * might make it difficult to see if the definition of `D` is located in an other file)
```

Comment:
The difficulty lies in understand why the predicate gives the results it does... Here is an other suggestion, let me know what you think!

```diff
- * inheritance chains (such as the result `C` for `cls=B` in the first example -- this
- * might make it difficult to see if the definition of `D` is located in an other file)
+ * inheritance chains: This is the case with the result `C` for `cls=B` in the first
+ * example -- if `B` and `C` are defined in the same file, but `D` in a different file,
+ * this might make the results from this predicate difficult to comprehend at first).
```

Reply:
Yes, I am happy with this one :-)

Reply:
Done in f262dc6